File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-8.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        int result = binarySearch(nums, target, 0, left - 1);
        if (result != -1) {
            return result;
        }

        return binarySearch(nums, target, left, nums.length - 1);
    }

    private int binarySearch(int[] nums, int target, int left, int right) {
        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target) 
                return mid;
            else if (nums[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }

        return -1;
    }
}

File: Data Structures & Algorithms/palindrome-partitioning/submission-8.cpp
class Solution {
    vector<vector<string>> result;

public:
    vector<vector<string>> partition(string s) {
        vector<string> list;
        recurse(s, 0, 0, list);

        return result;
    }

    void recurse(string &s, int j, int i, vector<string> &list) {
        if(i >= s.size()) {
            if(i == j) {
                result.push_back(list);
            }
            return;
        }

        if(isPalindrome(s, j, i)) {
            list.push_back(s.substr(j, i - j + 1));
            recurse(s, i + 1, i + 1, list);
            list.pop_back();
        }

        recurse(s, j, i + 1, list);
    }

    bool isPalindrome(string &s, int left, int right) {
        while(left < right) {
            if(s[left++] != s[right--]) return false;
        }

        return true;
    }
};
